core/domain/question_services.py

```diff
@@ -0,0 +1,205 @@
+
+
+def get_question_from_model(question_model):
+    """Returns domain object repersenting the given question model.
+
+    Args:
+        question_model: QuestionModel. The question model loaded from the
+            datastore.
+
+    Returns:
+        Question. The domain object representing the question model.
+    """
+    return question_domain.Question(
+        question_model.id, question_model.title, question_model.question_data,
+        question_model.question_data_schema_version,
+        question_model.collection_id, question_model.language_code)
+
+
+def get_question_by_id(question_id):
+    """Returns a domain object representing a question.
+
+    Args:
+        question_id: str. ID of the question.
+
+    Returns:
+        Question or None. The domain object representing a question with the
+        given id, or None if it does not exist.
+    """
+    question_model = question_models.QuestionModel.get(question_id)
+    if question_model:
+        question = get_question_from_model(question_model)
+        return question
+    else:
+        return None
+
+
+def apply_change_list(question_id, change_list):
+    """Applies a changelist to a pristine question and returns the result.
+
+    Args:
+        question_id: str. ID of the given question.
+        change_list: list(dict). A change list to be applied to the given
+            question. Each entry in change_list is a dict that represents an
+            QuestionChange object.
+
+    Returns:
+      Question. The resulting question domain object.
+    """
+    question = get_question_by_id(question_id)
+    try:
+        changes = [question_domain.QuestionChange(change_dict)
+                   for change_dict in change_list]
+
+        for change in changes:
+            if change.cmd == question_domain.CMD_EDIT_QUESTION_PROPERTY:
+                if (change.property_name ==
+                        question_domain.QUESTION_PROPERTY_TITLE):
+                    question.update_title(change.new_value)
+                elif (change.property_name ==
+                      question_domain.QUESTION_PROPERTY_LANGUAGE_CODE):
+                    question.update_language_code(change.new_value)
+                elif (change.cmd ==
+                      question_domain.QUESTION_PROPERTY_QUESTION_DATA):
+                    question.update_question_data(change.new_value)
+        return question
+
+    except Exception as e:
+        logging.error(
+            '%s %s %s %s' % (
+                e.__class__.__name__, e, question_id, change_list)
+        )
+        raise
+
+
+def _save_collection(committer_id, question, commit_message, change_list):
+    """Validates a question and commits it to persistent storage.
+
+    Args:
+        committer_id: str. The id of the user who is performing the update
+            action.
+        question: Question. The domain object representing a question.
+        change_list: list of dicts, each representing a QuestionChange object.
```

Comment:
Please follow existing format, and use actual domain objects in the list.

Reply:
But in collection_services, the `change_list` is a list of dicts, where each dict represents a CollectionChange.

Reply:
Yes, but my original comment still stands. We are introducing a new pattern here to use domain objects throughout.

Reply:
Updated.